### ISSUE TITLE
SslHandler doesn't support a use case where the input ByteBuf is shared across multiple threads

### ISSUE THREAD

**Issue by lhotari** [netty/netty#14069]
### Expected behavior

There should be a way to pass shared ByteBuf instances to a ChannelPipeline that contains SslHandler. A ByteBuf instance might be cached and that's why it's shared. 

An alternative is to document this limitation clearly in SslHandler javadoc.

### Actual behavior

Currently this will cause problems such as https://github.com/apache/pulsar/issues/2401 reported in Pulsar in 2018.
This was mitigated by calling `.copy()` on the input ByteBuf instances with https://github.com/apache/pulsar/pull/2464 . However, this `.copy()` solution had another issue which is reported as #14068. That is mitigated in Pulsar with https://github.com/apache/pulsar/pull/22760 . 

### Steps to reproduce

It should be possible to reproduce this issue by concurrently passing a ByteBuf instance to multiple channel pipelines that contain a SslHandler.

### Minimal yet complete reproducer code (or URL to code)

System level repro case for Pulsar is documented in https://github.com/apache/pulsar/issues/22601#issuecomment-2122642708 . This is not minimal.

### Netty version

4.1.108

### JVM version (e.g. `java -version`)

openjdk version "17.0.11" 2024-04-16
OpenJDK Runtime Environment Temurin-17.0.11+9 (build 17.0.11+9)
OpenJDK 64-Bit Server VM Temurin-17.0.11+9 (build 17.0.11+9, mixed mode, sharing)

### OS version (e.g. `uname -a`)

Linux pop-os 6.8.0-76060800daily20240311-generic #202403110203171407766522.04~4c8e9a0 SMP PREEMPT_DYNAMIC Thu A x86_64 x86_64 x86_64 GNU/Linux

**Comment by normanmaurer** [comment]
Passing the same `ByteBuf` instance to different `ChannelPipeline` is not supported at all. This is not really a limitation of the `SslHandler` but just the fact that `ByteBuf` is not thread-safe at all and once you pass a `ByteBuf` its ownership transfers.

**Comment by lhotari** [comment]
> Passing the same `ByteBuf` instance to different `ChannelPipeline` is not supported at all. This is not really a limitation of the `SslHandler` but just the fact that `ByteBuf` is not thread-safe at all and once you pass a `ByteBuf` its ownership transfers.

Understood. I guess the main problem is that it's hard to find where this is documented. 
Many Netty users might have the assumption that you could use `.retainedDuplicate()` in cases where ownership is transfered. Is this supported for this type of use cases?
btw. In Pulsar, `.copy()` was used together with SslHandler, but that has another problem, #14068. [The current WIP PR in Pulsar](https://github.com/apache/pulsar/pull/22760/files) replaces the use of `.copy()` with `.nioBuffers()/.nioBuffer()` to make a copy.
Would the correct solution be to use `.retainedDuplicate()` in the use case where there's a cached `ByteBuf` that is sent out to multiple clients?

**Comment by lhotari** [comment]
Checked quickly if `.retainedDuplicate()` could handle the ownership transfer, but it's not consistent. For pooled bytebufs, it seems that this code resolves the problem: 
https://github.com/netty/netty/blob/0dc95efd332fee260e9886cfbf66dd1d326da92f/buffer/src/main/java/io/netty/buffer/AbstractPooledDerivedByteBuf.java#L137-L139
However, for unpooled bytebufs, there will be issues when sharing across threads. 

It would be great if some API would guarantee a shallow copy which is safe for sharing. The mitigation in Pulsar PR apache/pulsar#22760 to use `.nioBuffers()/.nioBuffer()` and wrapping that into a ByteBuf with `Unpooled.wrappedBuffer` does the job, but it feels like a hack.

**Comment by lhotari** [comment]
I guess that `.retainedDuplicate()` should be sufficient when ownership is transfered?

Based on code in
https://github.com/netty/netty/blob/e5951d46fc89db507ba7d2968d2ede26378f0b04/buffer/src/main/java/io/netty/buffer/AbstractDerivedByteBuf.java#L116-L118
and
https://github.com/netty/netty/blob/0dc95efd332fee260e9886cfbf66dd1d326da92f/buffer/src/main/java/io/netty/buffer/AbstractPooledDerivedByteBuf.java#L137-L139

**Comment by lhotari** [comment]
In Pulsar, the `.copy()` solution was introduced in 2018 to fix another problem with Netty 4.1.22 which was used at that time.

There doesn't seem to be anything in SslHandler that would be causing the problems so I think that this issue that I have now reported here isn't a valid issue.

Any recommendations for sufficient ownership transfer to SslHandler will be appreciated since I'd like to make changes to Apache Pulsar (apache/pulsar#22760) based on Netty maintainer's recommendation.

**Comment by lhotari** [comment]
I've been trying to find out the answer to my own question. It looks like a read-only buffer is needed with SslHandler for the use case where a cached ByteBuf is sent to multiple clients. It would be useful to document this in the SslHandler javadoc. 
If the input ByteBuf isn't read-only it looks like SslHandlerCoalescingBufferQueue would attempt to concatenate the buffers when possible in [attemptCopyToCumulation](https://github.com/netty/netty/blob/32a025b4eb2c9fefb389ecd5c781f7cabc6ab139/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L2436-L2451).

**Comment by lhotari** [comment]
When I pass a read-only buffer, it logs a warning. I can file this as a separate issue if that makes it easier to track. It seems that the attemptCopyToCumulation method should check if the buffer is readonly before attempting to write to it.

```
2024-05-22T18:11:41,678+0000 [pulsar-io-3-4] WARN  org.apache.pulsar.broker.service.ServerCnx - [/10.34.1.241:60688] Got exception java.nio.ReadOnlyBufferException
   at io.netty.buffer.ReadOnlyByteBufferBuf.setBytes(ReadOnlyByteBufferBuf.java:384)
   at io.netty.buffer.DuplicatedByteBuf.setBytes(DuplicatedByteBuf.java:343)
   at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1104)
   at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1096)
   at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1087)
   at io.netty.handler.ssl.SslHandler.attemptCopyToCumulation(SslHandler.java:2446)
   at io.netty.handler.ssl.SslHandler.access$2800(SslHandler.java:171)
   at io.netty.handler.ssl.SslHandler$SslHandlerCoalescingBufferQueue.compose(SslHandler.java:2406)
   at io.netty.channel.AbstractCoalescingBufferQueue.remove(AbstractCoalescingBufferQueue.java:189)
   at io.netty.handler.ssl.SslHandler.wrap(SslHandler.java:841)
   at io.netty.handler.ssl.SslHandler.wrapAndFlush(SslHandler.java:822)
   at io.netty.handler.ssl.SslHandler.flush(SslHandler.java:803)
   at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:941)
   at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:921)
   at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:907)
   at io.netty.channel.ChannelOutboundHandlerAdapter.flush(ChannelOutboundHandlerAdapter.java:125)
   at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:939)
   at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:957)
   at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:982)
   at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:950)
   at org.apache.pulsar.broker.service.PulsarCommandSenderImpl.lambda$sendMessagesToConsumer$1(PulsarCommandSenderImpl.java:286)
   at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
   at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
   at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
   at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:413)
   at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
   at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
   at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
   at java.base/java.lang.Thread.run(Thread.java:840)
```

**Comment by normanmaurer** [comment]
> When I pass a read-only buffer, it logs a warning. I can file this as a separate issue if that makes it easier to track. It seems that the attemptCopyToCumulation method should check if the buffer is readonly before attempting to write to it.
> 
> ```
> 2024-05-22T18:11:41,678+0000 [pulsar-io-3-4] WARN  org.apache.pulsar.broker.service.ServerCnx - [/10.34.1.241:60688] Got exception java.nio.ReadOnlyBufferException
>    at io.netty.buffer.ReadOnlyByteBufferBuf.setBytes(ReadOnlyByteBufferBuf.java:384)
>    at io.netty.buffer.DuplicatedByteBuf.setBytes(DuplicatedByteBuf.java:343)
>    at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1104)
>    at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1096)
>    at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1087)
>    at io.netty.handler.ssl.SslHandler.attemptCopyToCumulation(SslHandler.java:2446)
>    at io.netty.handler.ssl.SslHandler.access$2800(SslHandler.java:171)
>    at io.netty.handler.ssl.SslHandler$SslHandlerCoalescingBufferQueue.compose(SslHandler.java:2406)
>    at io.netty.channel.AbstractCoalescingBufferQueue.remove(AbstractCoalescingBufferQueue.java:189)
>    at io.netty.handler.ssl.SslHandler.wrap(SslHandler.java:841)
>    at io.netty.handler.ssl.SslHandler.wrapAndFlush(SslHandler.java:822)
>    at io.netty.handler.ssl.SslHandler.flush(SslHandler.java:803)
>    at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:941)
>    at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:921)
>    at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:907)
>    at io.netty.channel.ChannelOutboundHandlerAdapter.flush(ChannelOutboundHandlerAdapter.java:125)
>    at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:939)
>    at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:957)
>    at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:982)
>    at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:950)
>    at org.apache.pulsar.broker.service.PulsarCommandSenderImpl.lambda$sendMessagesToConsumer$1(PulsarCommandSenderImpl.java:286)
>    at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
>    at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
>    at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
>    at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:413)
>    at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
>    at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
>    at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
>    at java.base/java.lang.Thread.run(Thread.java:840)
> ```

Yes please! Or do a PR :)

**Comment by normanmaurer** [comment]
> In Pulsar, the `.copy()` solution was introduced in 2018 with [apache/pulsar#2464](https://github.com/apache/pulsar/pull/2464) to fix another problem ([apache/pulsar#2401](https://github.com/apache/pulsar/issues/2401)) with Netty 4.1.22 which was used at that time.
> 
> There doesn't seem to be anything in SslHandler that would be causing the problems so I think that this issue that I have now reported here isn't a valid issue.
> 
> Any recommendations for sufficient ownership transfer to SslHandler will be appreciated since I'd like to make changes to Apache Pulsar ([apache/pulsar#22760](https://github.com/apache/pulsar/pull/22760)) based on Netty maintainer's recommendation. This issue can be closed. (I'll leave this open until someone replies. Thanks!)

I will review in detail and come back to you

**Comment by lhotari** [comment]
> Yes please! Or do a PR :)

@normanmaurer Here's the draft PR: #14071 . No unit tests yet.

**Comment by lhotari** [comment]
> Note that the JDK SSLEngine has had issues with read-only buffers in the past, because it likes to do encryption/decryption in-place. I don't think this is a well-trodden path.

@chrisvest Isn't that already handled with the `MERGE_CUMULATOR`? https://github.com/netty/netty/blob/7d51e908e6146fba59269d567618949ee1b11899/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L304
https://github.com/netty/netty/blob/a09cc1620c9f8ac0cb17851b85a051459ee02964/codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java#L98

If not, I guess it should be possible to add more changes for handling read-only buffers in [`SslHandlerCoalescingBufferQueue`](https://github.com/netty/netty/blob/7d51e908e6146fba59269d567618949ee1b11899/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L2388-L2428)?

**Comment by lhotari** [comment]
@normanmaurer there's another obstacle in supporting this use case where where a cached ByteBuf is sent to multiple clients use SslHandler. I believe that this is a very common and generic use case and it exists in Apache Pulsar. 
The solution idea was to first call `.asReadOnly()` for the original source ByteBuf before adding it to the cache.
However, this won't work since the read-only buffer doesn't support reference counting in the same way as a `.duplicate()` buffer. I've created another issue #14075 about this.

**Comment by lhotari** [comment]
@chrisvest Thanks for mentioning the JDK SSLEngine read-only buffer issue. I now found your email https://mail.openjdk.org/pipermail/security-dev/2022-March/029478.html 
and the related JDK bug reports https://bugs.openjdk.org/browse/JDK-8283577 and https://bugs.openjdk.org/browse/JDK-8285603. It seems that this is fixed [since Java jdk-19+23](https://github.com/openjdk/jdk/commit/f17c68ce4a0b4f5c3131f4e4626a5a55b7f2f61f).

**Comment by lhotari** [comment]
@chrisvest I started an early draft PR #14079 to address the JDK SSLEngine read-only buffer issue. Will continue next week based on feedback.

**Comment by lhotari** [comment]
@chrisvest @normanmaurer I have create a PR #14079 to address the JDK SSLEngine read-only buffer issue. It seems that after that is included, there would be support for the use case that this issue was about, supporting a use case where a cached ByteBuf is sent to multiple clients using SslHandler. I'd like to start using this solution in Apache Pulsar to replace the former solution where we always make a copy of all input when TLS is enabled. 
The idea will be to make the cached buffer read-only and that should be the way to signal SslHandler that it should make a copy of the input buffer if the SslEngineType wants a mutable buffer.

**Comment by lhotari** [comment]
Besides #14079 these are the other PRs that are needed: #14072, #14071, #14076 and #14078 (already merged to 4.1 branch).

**Comment by lhotari** [comment]
I checked the Netty code once again and noticed that it's possible to optimize the current implementation in SslHandler and avoid the buffer mutation. I created #14086. Please review @normanmaurer @chrisvest. Thanks!
